Src/PerformanceCollector/Shared/QuickPulseTelemetryModule.cs

```diff
@@ -0,0 +1,468 @@
+        {
+            return configuration.TelemetryProcessors.OfType<IQuickPulseTelemetryProcessor>().SingleOrDefault();
+        }
+
+        private void InitializeServiceClient(TelemetryConfiguration configuration)
+        {
+            if (this.serviceClient != null)
+            {
+                // service client has been passed through a constructor, we don't need to do anything
+                return;
+            }
+
+            Uri serviceEndpointUri;
+            if (string.IsNullOrWhiteSpace(this.QuickPulseServiceEndpoint))
+            {
+                // endpoint is not specified in configuration, use the default one
+                serviceEndpointUri = this.serviceUriDefault;
+            }
+            else
+            {
+                // endpoint appears to have been specified in configuration, try using it
+                try
+                {
+                    serviceEndpointUri = new Uri(this.QuickPulseServiceEndpoint);
+                }
+                catch (Exception e)
+                {
+                    throw new ArgumentException(
+                        string.Format(
+                            CultureInfo.InvariantCulture,
+                            "Error initializing QuickPulse module. QPS endpoint is not a correct URI: '{0}'",
+                            this.QuickPulseServiceEndpoint),
+                        e);
+                }
+            }
+
+            // create the default production implementation of the service client with the best service endpoint we could get
+            this.serviceClient = new QuickPulseServiceClient(serviceEndpointUri, GetInstanceName(configuration), SdkVersionUtils.GetAssemblyVersion());
+        }
+
+        private static string GetInstanceName(TelemetryConfiguration configuration)
+        {
+            // we need to initialize an item to get instance information
+            var fakeItem = new MetricTelemetry();
+
+            try
+            {
+                new TelemetryClient(configuration).Initialize(fakeItem);
+            }
+            catch (Exception)
+            {
+                // we don't care what happened there
+            }
+
+            return string.IsNullOrWhiteSpace(fakeItem.Context?.Cloud?.RoleInstance) ? Environment.MachineName : fakeItem.Context.Cloud.RoleInstance;
+        }
+        
+        private void StateTimerCallback(object state)
+        {
+            var currentCallbackStarted = this.timeProvider.UtcNow;
+            TimeSpan? timeToNextUpdate = null;
+
+            try
+            {
+                // the first instrumentation key that we get from TelemetryConfiguration.Active
+                // will be our permanent instrumentation key
+                if (string.IsNullOrWhiteSpace(this.instrumentationKey))
+                {
+                    this.instrumentationKey = TelemetryConfiguration.Active.InstrumentationKey;
```

Comment:
Use one that passed in `Initialize`. Also many people initialize instrumentation key in Global.asax after all modules already ininitalized